### DISCUSSION TITLE
[Chapter 08] Spring Data JPA

### DISCUSSION THREAD

**Discussion by coalong** [cowaine/spring-boot-for-msa#95] [Chapter 08. Data Persistence] [answered]
P. 457 쪽에 보면 아래 설명이 나와있다.

> JPA 프레임워크에 대한 높은 이해도: 
JPA는 객체와 관계형 데이터베이스를 매핑하는 프레임워크로 개발자는 Hibernate나 Spring Data JPA의 메서드를 사용하여 개발할 수 있다. 
그러면 Hibernate 내부에서는 매핑 정보를 사용하여 쿼리를 생성한다. 한 단계 추상화된 레이어이므로 직접 쿼리를 작성하여 실행하는 것보다 성능상 단점이 있다. 
Hibernate에서는 성능 향상을 위한 여러 방법을 제공하고 있으며, 개발자는 이에 대한 이해가 필요하다.

`직접 쿼리를 사용하는 것보다 추상화된 ORM 을 사용하여 쿼리를 실행하는 건 성능상 단점이 있고, 성능 향상을 위해 여러 방법을 제공하고 있다고 한다.`

직접 쿼리를 실행하는 것보다 성능상 차이가 많을까? 
혹시 ORM를 사용하여 성능 저하를 겪어본 적이 있는지, 또한 성능 향상을 위해 여러 방법을 사용해본 경험이 있거나 알고있는 성능 향상 방법이 뭐가 있을지 이야기 해보고싶다!

**Comment by ahngilwoong** [comment, marked as answer]
예전에 학습했던 JPA의 N+1 문제도 성능저하 예시중의 하나일 것 같습니다.
[N+1 문제](https://velog.io/@sweet_sumin/JPA-N1-%EC%9D%B4%EC%8A%88%EB%8A%94-%EB%AC%B4%EC%97%87%EC%9D%B4%EA%B3%A0-%ED%95%B4%EA%B2%B0%EC%B1%85%EC%9D%80-%EB%AC%B4%EC%97%87%EC%9D%B8%EA%B0%80%EC%9A%94)
실제로 제가 담당하는 MES에서는 JPA를 사용하는데 엄청나게 복잡한 쿼리 성능이 중요한 쿼리들은 `@Query`를 사용하거나 Mybatis를 사용해버리는 것으로 확인했습니다(조회가 엄청 많고 조회 성능이 중요한 서버에 한정) 
이 방법은 뭔가 JPA내에서 기술적으로 풀어 나간 것은 아니지만 하나의 방법이 될 수도 있다고 생각합니다.

**Comment by Irisation23** [comment]
### 질의 응답 이외의 CheckPoint

1. 만약 `Dynamic Update` `Bulk Insert`에 대해서는 어떻게 생각할 수 있을까?
2. `Transaction의 readOnly=XXX` 적용이 성능에 미치는 영향
